### PR TITLE
feat(ldap): sync additional properties to profile and SAB

### LINKED ISSUE #40194: More LDAP User Profile Attributes
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are interested into the same feature.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

**Is your feature request related to a problem? Please describe.**
No.

**Describe the solution you'd like**
It would be fantastic if you could integrate more LDAP user profile attributes, such as birthday or timezone. This would make it easier to fully integrate LDAP user management with Nextcloud Contacts. Ideally any contact card field can be passed from LDAP. 

**Describe alternatives you've considered**
None, this only concerns LDAP and Nextcloud integration.

**Additional context**
None.

### PR DESCRIPTION
**Pull request by st3iny**
* Resolves #40194
* Continuation of #41793

## Summary

This PR merges #41793 and #39592. I added neccessary frontend code so that users can change their date of births and adjust the scope in case they don't want it to be synced to the SAB.

I also removed the anniversary date after having a discussion with our designers. We are required to take the detour via the profile page to allow users to set the scope. Considering the broad spectrum that Nextcloud is used in, it doesn't make sense to have an anniversary date on the profile.

Synced from LDAP to profile:
- Date of birth

Synced from LDAP to SAB (via the profile):
- Biography
- Date of birth

Original code by Jake Nabasny (GitHub: @slapcat). Your contribution is highly valued :heart: 

### Synced contact in SAB
Bildschirmfoto_20240526_105524

### Date of birth in personal settings
Bildschirmfoto_20240526_110044

### LDAP wizard (advanced tab)
Bildschirmfoto_20240526_105358

## TODO

- [x] Tests (probably)

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Comment by st3iny** [comment]
Pending documentation at https://github.com/nextcloud/documentation/pull/11323 (although it has to be adjusted to the removed anniversary field).

**Review by come-nc — COMMENTED** [review]
What kind of LDAP attribute was this tested with?
I’m afraid it won’t work with LDAP date format which is quite specific and not supported by PHP out-of-the-box.

**Comment by st3iny** [comment]
Côme made a good point. There is no standardized birthdate field on LDAP. However, some implementations use the standardized generalized date format while others user the `YYYY-MM-DD` format. 

So we need a parser for the LDAP format and fallback to the date constructor.

**Comment by slapcat** [comment]
@come-nc I tested it with an attribute using the same Generalized Time syntax that you linked and it worked fine for both showing the date in the frontend and generating it on contact cards for the cardDAV server. Here is the full attribute type I use:
```
olcAttributeTypes: {0}( 1.3.6.1.4.1.59010.1.1 NAME 'birthDate' DESC 'Birthday'
  EQUALITY generalizedTimeMatch ORDERING generalizedTimeOrderingMatch SYNTAX 1
 .3.6.1.4.1.1466.115.121.1.24 SINGLE-VALUE )
```

**Comment by come-nc** [comment]
> @come-nc I tested it with an attribute using the same Generalized Time syntax that you linked and it worked fine for both showing the date in the frontend and generating it on contact cards for the cardDAV server. Here is the full attribute type I use:
> 
> ```
> olcAttributeTypes: {0}( 1.3.6.1.4.1.59010.1.1 NAME 'birthDate' DESC 'Birthday'
>   EQUALITY generalizedTimeMatch ORDERING generalizedTimeOrderingMatch SYNTAX 1
>  .3.6.1.4.1.1466.115.121.1.24 SINGLE-VALUE )
> ```

Using the two examples from the RFC the second one crashes DateTime constructor:
https://3v4l.org/uSuTs

Try setting your attribute to "199412160532-0500" which is a valid value from the RFC.

**Comment by st3iny** [comment]
I added a rudimentary parser, that just uses the first 8 characters from the LDAP generalized syntax and tries to parse YYYYMMDD or YYYY-MM-DD otherwise. Is that sufficient?

I also tested it with a real LDAP generalized time attribute on my local dev instance. Of course, I also added some unit test for the parser method ...

File: apps/settings/src/components/PersonalInfo/BirthdaySection.vue [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+		if (birthdate.value) {
+			initialValue = new Date(birthdate.value)
+		}
+
+		return {
+			birthdate: {
+				...birthdate,
+				readable: NAME_READABLE_ENUM[birthdate.name],
+			},
+			initialValue: initialValue ? new Date(initialValue) : new Date(),
```
Comment (ChristophWurst) [inline apps/settings/src/components/PersonalInfo/BirthdaySection.vue:60]:
```suggestion
			initialValue: initialValue ? initialValue : new Date(),
```

it already is a date object, right?

File: apps/user_ldap/lib/Service/BirthdateParserService.php [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+class BirthdateParserService {
+	/**
+	 * Try to parse the birthdate from LDAP.
+	 * Supports LDAP's generalized time syntax, YYYYMMDD and YYYY-MM-DD.
+	 *
+	 * @throws InvalidArgumentException If the format of then given date is unknown
+	 */
+	public function parseBirthdate(string $value): DateTimeImmutable {
+		if (strlen($value) >= 11) {
```
Comment (ChristophWurst) [inline apps/user_ldap/lib/Service/BirthdateParserService.php:22]:
```suggestion
		if (strlen($value) >= strlen('YYYY-MM-DD') {
```

nit: no magic numbers

**Review by ChristophWurst — APPROVED** [review]
Looks good

some bundle updates are suspicious. e.g. dist/weather_status-weather-status.js changed

File: apps/settings/src/components/PersonalInfo/BirthdaySection.vue (hunk 1 above)
Comment (st3iny) [reply apps/settings/src/components/PersonalInfo/BirthdaySection.vue:60]: Good catch! Done

File: apps/user_ldap/lib/Service/BirthdateParserService.php (hunk 2 above)
Comment (st3iny) [reply apps/user_ldap/lib/Service/BirthdateParserService.php:22]: Good point. I also added a comment about the formats.

**Review by ChristophWurst — APPROVED** [review]

File: apps/user_ldap/lib/Service/BirthdateParserService.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	 *
+	 * @throws InvalidArgumentException If the format of then given date is unknown
+	 */
+	public function parseBirthdate(string $value): DateTimeImmutable {
+		// Minimum LDAP generalized date is "1994121610Z" with 11 chars
+		// While maximum other format is "1994-12-16" with 10 chars
+		if (strlen($value) > strlen('YYYY-MM-DD')) {
+			// Probably LDAP generalized time syntax
+			$value = substr($value, 0, 8);
+		}
```
Comment (come-nc) [inline apps/user_ldap/lib/Service/BirthdateParserService.php:27]: Hum that will break on ISO format like "2004-02-12T15:19:21+00:00", which may be out there as well.

Maybe check that first 10 characters are digits before `substr`?

File: apps/user_ldap/lib/Service/BirthdateParserService.php [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+		// Should be either YYYYMMDD or YYYY-MM-DD
+		if (!preg_match('/^(\d{8}|\d{4}-\d{2}-\d{2})$/', $value)) {
+			throw new InvalidArgumentException("Unknown date format: $value");
+		}
```
Comment (come-nc) [inline apps/user_ldap/lib/Service/BirthdateParserService.php:32]: Ah I missed that. I would use a more flexible converter, if DateTime does not choke on it we accept it (apart from the special handling for generalizedtime).

**Review by come-nc — APPROVED** [review]

**Comment by szaimen** [comment]
🎉🎉🎉🎉🎉

File: lib/public/Accounts/IAccountManager.php [hunk 5]
```diff
@@ -156,6 +156,11 @@ interface IAccountManager {
 	 */
 	public const PROPERTY_PROFILE_ENABLED = 'profile_enabled';
 
+	/**
+	 * @since 30.0.0
+	 */
+	public const PROPERTY_BIRTHDATE = 'birthdate';
```
Comment (nickvergessen) [inline lib/public/Accounts/IAccountManager.php:162]: Please document in Please document in https://docs.nextcloud.com/server/latest/developer_manual/app_publishing_maintenance/app_upgrade_guide/upgrade_to_30.html#id1

**Comment by aurisnoctis** [comment]
Will this change allow people to add their birthday to their profile even if the nextcloud server uses not LDAP but the default database?

**Comment by st3iny** [comment]
@aurisnoctis Yes, it does work without LDAP.

**Pull request merged**